docs/internal/versioning.md

```diff
@@ -0,0 +1,244 @@
+API versions use semantic versioning, where the major version is incremented to indicate the removal of types or methods and the minor version to indicate the addition of new classes or methods.
+Both indicate potentially breaking changes for Verona consumers but it should be possible for a package to support all minor versions of an API with minimal effort by the package author.
+There is also one special API version: `unstable`.
+This represents any work-in-progress changes since the last release and exists so that developers can import the current snapshot of a library to test whether in-development changes are likely to cause problems.
+
+By making the API version distinct from the implementation version, we allow:
+
+ - A program to consume two different API versions of the same package (while consuming a single implementation)
+ - A package to deprecate and remove APIs gradually over time without needing two versions of the code.
+ - A package to add new APIs without breaking any consumers.
+
+When a package is imported, the string describing the package should include the API version, for example:
+
+```verona
+import "some/package@2.0" as p;
+```
+
+The build tooling should pick the most recent version of the package that supports all of the versions that are used in the project and its dependencies and warn if that is not the most recent available version.
+**Open question: Should pinning to specific implementations be part of this or should it be separate build metadata such as vcpkg.json?**
+
+When the package is imported with this version, every class, interface, method, or field that is declared as public but is *not* exposed by this version is implicitly package-private.
+This means that it cannot name any type or field.
+
+Implementation
+--------------
+
+Package-private interface and class types are simply not nameable and therefore not usable in other packages.
+The same is true if they are implicitly hidden by belonging to a version that is not in use.
+This is purely a compile-time check.
+
+Methods and fields are slightly more complex.
+Methods or fields accessed on concrete types are trivially allowed or disallowed depending on whether the caller can see an API version that includes them.
+Method or field accesses via interfaces depend on whether the selector used in the calling package unifies with the versioned selector in the callee's package.
+
+Package-private members effectively introduce an anonymous (or, at least, unnameable) namespace for selectors such that `_foo` in one package and `_foo` in another are different selectors and there is no way of code in one package naming the other selector.
+Versions simply generalise this slightly.
+Versioned members are identified by a selector that includes the version and package of origin, which has the same value as a selector of the same name and arrity in a package that imports them and is permitted to see that version.
```

Comment:
This will probably require some mangling rules, which each different mangled symbol pointing to a different selector.

Reply:
It's not just mangling, it will also require some unification rules to define *which* mangling each package sees for unversioned selectors.